File: src/diceRoller/utils/parseDiceUserInput.ts
import { Dice } from "../types";

const ALLOWED_DICE_SIDES = new Set(Object.values(Dice));

export function parseDiceUserInput(input: string): DiceParseResult {
	if (typeof input !== "string" || !input.trim()) {
		throw new Error("Input must be a non-empty string.");
	}

	let clean = input.toLowerCase().trim();
	clean = clean.replace(/\b(roll|dice|die)\b/g, "");
	clean = clean.replace(/\s+/g, "");

	if (/^\d+$/.test(clean)) {
		const sides = Number.parseInt(clean, 10);
		if (!ALLOWED_DICE_SIDES.has(sides)) {
			throw new Error(
				`Allowed dice sides are ${[...ALLOWED_DICE_SIDES].join(", ")}. Received: ${sides}`,
			);
		}
		return { dices: [sides], mod: 0 };
	}

	const tokenRegex = /(\d*d\d+|[+\-]\d+)/g;
	const tokens = clean.match(tokenRegex);

	if (!tokens) {
		throw new Error(
			"Unable to parse any dice or modifiers. Examples: '3d6+5', 'd20-2', '2d8+1d6+4'.",
		);
	}

	const dices: number[] = [];
	let mod = 0;

	for (const token of tokens) {
		if (token.includes("d")) {
			const [countPart, sidesPart] = token.split("d");

			let diceCount = countPart ? Number.parseInt(countPart, 10) : 1;
			if (Number.isNaN(diceCount) || diceCount < 1) {
				diceCount = 1;
			}

			const sides = Number.parseInt(sidesPart, 10);
			if (Number.isNaN(sides) || sides <= 0) {
				throw new Error(`Invalid number of sides: "${sidesPart}"`);
			}

			if (!ALLOWED_DICE_SIDES.has(sides)) {
				throw new Error(
					`Allowed dice sides are ${[...ALLOWED_DICE_SIDES].join(", ")}. Received: ${sides}`,
				);
			}

			for (let i = 0; i < diceCount; i++) {
				dices.push(sides);
			}
		} else {
			const parsedMod = Number.parseInt(token, 10);
			if (Number.isNaN(parsedMod)) {
				throw new Error(`Invalid modifier: "${token}"`);
			}
			mod += parsedMod;
		}
	}

	return { dices, mod };
}

File: DiscordBots.BoredBot/DiceRoller/Utils/ParseDiceUserInput.cs
using System.Text.RegularExpressions;

namespace DiscordBots.BoredBot.DiceRoller.Utils
{
    public static class ParseDiceUserInput
    {
        private static readonly HashSet<int> AllowedDiceSides = new()
        {
            Dice.Four, Dice.Six, Dice.Ten, Dice.Twelve, Dice.Twenty, Dice.Hundred
        };

        public static DiceParseResult Parse(string input)
        {
            if (string.IsNullOrWhiteSpace(input))
            {
                throw new ArgumentException("Input must be a non-empty string.");
            }

            var clean = input.ToLowerInvariant().Trim();
            clean = Regex.Replace(clean, @"\b(roll|dice|die)\b", "");
            clean = Regex.Replace(clean, @"\s+", "");

            if (Regex.IsMatch(clean, @"^\d+$"))
            {
                var sides = int.Parse(clean);
                if (!AllowedDiceSides.Contains(sides))
                {
                    throw new ArgumentException($"Allowed dice sides are {string.Join(", ", AllowedDiceSides)}. Received: {sides}");
                }
                return new DiceParseResult { Dices = new List<int> { sides }, Mod = 0 };
            }

            var tokenRegex = new Regex(@"(\d*d\d+|[+\-]\d+)");
            var tokens = tokenRegex.Matches(clean).Cast<Match>().Select(m => m.Value).ToArray();

            if (tokens.Length == 0)
            {
                throw new ArgumentException("Unable to parse any dice or modifiers. Examples: '3d6+5', 'd20-2', '2d8+1d6+4'.");
            }

            var dices = new List<int>();
            var mod = 0;

            foreach (var token in tokens)
            {
                if (token.Contains("d"))
                {
                    var parts = token.Split('d');
                    var countPart = parts[0];
                    var sidesPart = parts[1];

                    var diceCount = string.IsNullOrEmpty(countPart) ? 1 : int.Parse(countPart);
                    if (diceCount < 1) diceCount = 1;

                    var sides = int.Parse(sidesPart);
                    if (sides <= 0)
                    {
                        throw new ArgumentException($"Invalid number of sides: \"{sidesPart}\"");
                    }

                    if (!AllowedDiceSides.Contains(sides))
                    {
                        throw new ArgumentException($"Allowed dice sides are {string.Join(", ", AllowedDiceSides)}. Received: {sides}");
                    }

                    for (int i = 0; i < diceCount; i++)
                    {
                        dices.Add(sides);
                    }
                }
                else
                {
                    var parsedMod = int.Parse(token);
                    mod += parsedMod;
                }
            }

            return new DiceParseResult { Dices = dices, Mod = mod };
        }
    }
}